hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/protocol/datatransfer/sasl/SaslDataTransferClient.java

```diff
@@ -603,7 +603,20 @@ private IOStreamPair doSaslHandshake(InetAddress addr,
           conf, cipherOption, underlyingOut, underlyingIn, false) :
           sasl.createStreamPair(out, in);
     } catch (IOException ioe) {
-      sendGenericSaslErrorMessage(out, ioe.getMessage());
+      String message = ioe.getMessage();
+      try {
+        sendGenericSaslErrorMessage(out, message);
+      } catch (Exception e) {
+        // If ioe is caused by error response from server, server will close peer connection.
+        // So sendGenericSaslErrorMessage might cause IOException due to "Broken pipe".
+        // We suppress IOException from sendGenericSaslErrorMessage
+        // and always throw `ioe` as top level.
+        // `ioe` can be InvalidEncryptionKeyException or InvalidBlockTokenException
+        // that indicates refresh key or token and are important for caller.
+        LOG.debug("Failed to send generic sasl error (server: {}, message: {}), suppress exception",
+            addr.toString(), message, e);
```

Comment:
Thank you for the update. In SLF4J API, `toString()` is not necessary. Passing the Object instead of String allows lazy-evaluation and it skips `toString()` method when the log level is not debug or trace.
```suggestion
            addr, message, e);
```

Reply:
Ah, indeed fixed with similar codes around here https://github.com/apache/hadoop/pull/3677/commits/495ba76857851a6acfc948e3b373f40449c2c873

Reply:
Thanks!